### PR TITLE
feat(quic): allow disabling of path MTU discovery

### PR DESCRIPTION
**Pull request by nazar-pc**
## Description

Path MTU discovery seems to be broken in QUIC right now on Windows and we had to disable it in Subspace.

Related: https://github.com/subspace/subspace/pull/2195.

## Attributions

Co-authored-by: Shamil Gadelshin <shamilgadelshin@gmail.com>

## Notes & open questions

It is following default behavior of quinn and just adds an option to change that if necessary.

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: transports/quic/src/config.rs [hunk 1]
```diff
@@ -83,8 +86,17 @@ impl Config {
             // Ensure that one stream is not consuming the whole connection.
             max_stream_data: 10_000_000,
             keypair: keypair.clone(),
+            path_mtu_discovery_config: Some(Default::default()),
         }
     }
+
+    pub fn path_mtu_discovery_config(
+        mut self,
+        path_mtu_discovery_config: Option<MtuDiscoveryConfig>,
```
Comment (thomaseizinger) [inline transports/quic/src/config.rs]: This adds `quinn` to the public API which means every dependency bump will be a breaking change.

If all we need for now is disabling path MTU discovery, can we add that as an option instead, i.e. `disable_path_mtu_discovery`.

Comment (nazar-pc) [reply transports/quic/src/config.rs]: The type can be re-exported, but it is still a potential source of breaking changes. Changed public method to `disable_path_mtu_discovery` as suggested.

File: transports/quic/src/config.rs [hunk 2]
```diff
@@ -63,6 +63,9 @@ pub struct Config {
     server_tls_config: Arc<rustls::ServerConfig>,
     /// Libp2p identity of the node.
     keypair: libp2p_identity::Keypair,
+
+    /// Parameters governing MTU discovery. See [`MtuDiscoveryConfig`] for details.
+    path_mtu_discovery_config: Option<MtuDiscoveryConfig>,
```
Comment (mxinden) [inline transports/quic/src/config.rs]:
```suggestion
    mtu_discovery_config: Option<MtuDiscoveryConfig>,
```

Nit pick to be consistent with the `quinn` type.

**Review by mxinden — COMMENTED** [review]
:+1: works for me. Can you add a changelog entry?

**Comment by thomaseizinger** [comment]
Note to myself: Tidy up PR description before merging this.

File: transports/quic/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,10 @@
+## 0.10.1 - unreleased
+
+- Allow disabling MTU discovering.
+  See [PR 4823].
+
+[PR 4823]: https://github.com/libp2p/rust-libp2p/pull/4823
```
Comment (thomaseizinger) [inline transports/quic/CHANGELOG.md]:
```suggestion
- Allow disabling path MTU discovery.
  See [PR 4823](https://github.com/libp2p/rust-libp2p/pull/4823).
```

File: transports/quic/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,10 @@
+## 0.10.1 - unreleased
```
Comment (thomaseizinger) [inline transports/quic/CHANGELOG.md:1]: Please also bump the version in the manifest (see CI error).

**Review by thomaseizinger — COMMENTED** [review]
Thanks! just one formality.

File: transports/quic/CHANGELOG.md (hunk 3 above)
Comment (nazar-pc) [reply transports/quic/CHANGELOG.md]: Wasn't sure which format to use, there were conflicting examples :slightly_smiling_face:

File: transports/quic/Cargo.toml [hunk 5]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "libp2p-quic"
-version = "0.10.0"
+version = "0.10.1"
```
Comment (thomaseizinger) [inline transports/quic/Cargo.toml:3]: Also needs a bump in the main `Cargo.toml`, sorry!

**Comment by nazar-pc** [comment]
Yep, apologies, this trivial PR will certainly need to be squashed

**Comment by thomaseizinger** [comment]
> Yep, apologies, this trivial PR will certainly need to be squashed

We always squash-merge! :)

**Review by thomaseizinger — APPROVED** [review]

**Pull request merged**